models/user.go

```diff
@@ -435,15 +435,43 @@ func FindUserByPhoneChangeAndAudience(tx *storage.Connection, phone, aud string)
 }
 
 // IsDuplicatedEmail returns whether a user exists with a matching email and audience.
-func IsDuplicatedEmail(tx *storage.Connection, email, aud string) (bool, error) {
-	_, err := FindUserByEmailAndAudience(tx, email, aud)
-	if err != nil {
-		if IsNotFoundError(err) {
-			return false, nil
+func IsDuplicatedEmail(tx *storage.Connection, email, aud string) (*User, error) {
+	var identities []Identity
+
+	if err := tx.Eager().Q().Where("email = ?", strings.ToLower(email)).All(&identities); err != nil {
+		if errors.Cause(err) == sql.ErrNoRows {
+			return nil, nil
 		}
-		return false, err
+
+		return nil, errors.Wrap(err, "unable to find identity by email for duplicates")
 	}
-	return true, nil
+
+	userIDs := make(map[string]uuid.UUID)
+	for _, identity := range identities {
+		if !identity.IsForSSOProvider() {
+			userIDs[identity.UserID.String()] = identity.UserID
+		}
+	}
+
+	for _, userID := range userIDs {
+		user, err := FindUserByID(tx, userID)
+		if err != nil && !IsNotFoundError(err) {
+			return nil, errors.Wrap(err, "unable to find user from email identity for duplicates")
+		}
+
+		if user.Aud == aud {
+			return user, nil
+		}
+	}
```

Comment:
i feel like this can be simplified to a `select distinct(user_id) from auth.identities where email = ? and left(provider, 3) != 'sso';`, wdyt?

Reply:
Yes but I'd rather not write raw queries unless absolutely necessary at this stage... readability is more important than performance in this case I think.

Reply:
hmm doesn't the raw query improve both readability _and_ performance though? you can avoid having to use the map as a set-replacement so it would look something like this:


```go
var userIDs []uuid.UUID
if err := tx.RawQuery("select distinct(user_id) from "+(&pop.Model{Value: Identity{}}).TableName()+" where email = ? and left(provider, 3) != 'sso'", email).All(&userIDs); err != nil {
  if errors.Cause(err) == sql.ErrNoRows {
	return nil, nil
  }
  return nil, errors.Wrap(err, "unable to find identity by email for duplicates")
}
for _, userID := range userIDs {
  user, err := FindUserByID(tx, userID)
  if err != nil && !IsNotFoundError(err) {
    return nil, errors.Wrap(err, "unable to find user from email identity for duplicates")
  }

  if user.Aud == aud {
    return user, nil
  }
}
```

Reply:
To me the reason why that is less readable is because the code for the SSO part is found within a string, and the intention is somewhat difficult to gauge on first read. The map/set problem is real though, but unlike one query there's less information packed per line... we can refactor this of course if you feel very strongly about it.

Reply:
yeah that's true, it's alright, seems fine for now!